File: studio/services/composition_service.py
"""Compose website variants from the extracted component library."""

import json
import logging
import os
import random
import re
import time

from studio.services import onboarding_service

_log = logging.getLogger(__name__)

_STUDIO_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
COMPONENTS_DIR = os.path.join(_STUDIO_DIR, "components")
TEMPLATES_DIR = os.path.join(_STUDIO_DIR, "templates")

RECIPE = [
    "nav",
    "hero",
    "features",
    "about",
    "gallery",
    "stats",
    "pricing",
    "testimonials",
    "faq",
    "cta",
    "contact",
    "footer",
]
REQUIRED = frozenset({"nav", "hero", "footer"})
_MAX_HTML_BYTES = 150 * 1024

_index_cache = None
_file_cache = {}

_SHELL = """<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>{{BRAND}}</title>
<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/css/bootstrap.min.css" rel="stylesheet">
__FONTS__
<style>__BASE__</style>
<style>__PALETTE__</style>
</head>
<body>
__BODY__
</body>
</html>
"""

_SELECT_PROMPT = """Pick the page design that best fits this small business.

Business genre: {genre}
Business:
{facts}

Page designs (JSON) — each is one complete, visually coherent page:
{designs}

Reply with only JSON naming ONE design, like {{"design": "2"}}.
Use only a design id listed above."""


class CompositionError(RuntimeError):
    """Raised when a variant cannot satisfy its recipe."""


def reset_for_tests():
    global _index_cache
    _index_cache = None
    _file_cache.clear()


def _index():
    global _index_cache
    if _index_cache is None:
        path = os.path.join(COMPONENTS_DIR, "manifest.json")
        with open(path, encoding="utf-8") as f:
            entries = json.load(f)
        idx = {}
        for entry in entries:
            idx.setdefault(entry["type"], []).append(entry)
        _index_cache = idx
    return _index_cache


def _read(name):
    if name not in _file_cache:
        with open(os.path.join(COMPONENTS_DIR, name), encoding="utf-8") as f:
            _file_cache[name] = f.read()
    return _file_cache[name]


def _family(entry):
    """Which source page a component was extracted from.

    Ids are {genre}-{page}-{index}-{type} for the multi-page sets and
    {genre}-{index}-{type} for the first page. Components from different pages
    carry different fonts, button shapes and section backgrounds, so mixing them
    produces a page that reads as several sites stapled together.
    """
    parts = entry["id"].split("-")
    return parts[1] if len(parts) == 4 else "0"


def _order(entry):
    """Original position on the source page — keeps the designer's section flow."""
    parts = entry["id"].split("-")
    return int(parts[2]) if len(parts) == 4 else int(parts[1])


def _families(genre):
    """{family: {slot: [entries]}} for families that can build a whole page."""
    fams = {}
    for slot in RECIPE:
        for entry in _index().get(slot, []):
            if entry["genre"] == genre:
                fams.setdefault(_family(entry), {}).setdefault(slot, []).append(entry)
    return {f: slots for f, slots in fams.items() if REQUIRED <= set(slots)}


_ROOT_RE = re.compile(r"(?s):root\s*\{[^}]*\}")


def _palette(entry):
    """The :root palette of the page this component was extracted from.

    base.css ships one palette for the whole library, so without this every
    genre renders in the same colours (saas green) and the "themes" the
    templates actually define never reach a composed page.
    """
    fam = _family(entry)
    name = f"{entry['genre']}.html" if fam == "0" else f"{entry['genre']}-{fam}.html"
    if name not in _file_cache:
        try:
            with open(os.path.join(TEMPLATES_DIR, name), encoding="utf-8") as f:
                found = _ROOT_RE.search(f.read())
        except OSError:
            found = None
        _file_cache[name] = found.group(0) if found else ""
    return _file_cache[name]


def _build(slots, rng):
    chosen = [rng.choice(slots[s]) for s in RECIPE if s in slots]
    return sorted(chosen, key=_order)


def _pick_deterministic(project_id, genre, seed):
    """One whole page family, so every section shares one visual language."""
    fams = _families(genre)
    if not fams:
        raise CompositionError(f"no complete {genre} component family")
    rng = random.Random(f"{project_id}:{seed}")
    return _build(fams[rng.choice(sorted(fams))], rng)


def _pick_llm(genre, business_facts, rng):
    """LLM picks the page family. None = use deterministic path."""
    provider, model = onboarding_service._load_onboarding_provider()
    if provider is None or not model:
        return None
    fams = _families(genre)
    if not fams:
        return None
    designs = {
        f: {
            "sections": [s for s in RECIPE if s in slots],
            "tags": sorted(
                {t for pool in slots.values() for e in pool for t in e["tags"]} - {genre}
            ),
        }
        for f, slots in fams.items()
    }
    prompt = _SELECT_PROMPT.format(
        genre=genre,
        facts=business_facts or "(no details provided)",
        designs=json.dumps(designs, indent=0),
    )
    try:
        out = provider.generate(model, prompt, num_ctx=8192, timeout_s=30)
    except Exception:
        _log.warning("llm selection call failed; using deterministic", exc_info=True)
        return None
    raw = re.sub(r"^```(?:json)?\s*|\s*```$", "", (out or "").strip())
    try:
        data = json.loads(raw)
    except ValueError:
        return None
    if not isinstance(data, dict):
        return None
    slots = fams.get(str(data.get("design")))
    if slots is None:
        return None
    return _build(slots, rng)


def _assemble(entries):
    fonts, seen, parts = [], set(), []
    for entry in entries:
        parts.append(_read(entry["path"]))
        for href in entry.get("fonts", []):
            if href not in seen:
                seen.add(href)
                fonts.append(f'<link href="{href}" rel="stylesheet">')
    html = (
        _SHELL.replace("__FONTS__", "\n".join(fonts))
        .replace("__BASE__", _read("base.css"))
        .replace("__PALETTE__", _palette(entries[0]))
        .replace("__BODY__", "\n".join(parts))
    )
    for slot in ("nav", "hero", "footer"):
        if f'data-section="{slot}"' not in html:
            raise CompositionError(f"assembled html missing data-section={slot}")
    return html


def compose_project_variants(project_id, business_facts, genre, count):
    """Compose up to `count` page variants. Never raises; failures are skipped."""
    if count <= 0:
        return []
    variants = []
    for seed in range(count):
        started = time.perf_counter()
        try:
            entries, via = None, "deterministic"
            if seed == 0:
                try:
                    entries = _pick_llm(
                        genre, business_facts, random.Random(f"{project_id}:llm")
                    )
                except Exception:
                    _log.exception("llm selection crashed; using deterministic")
                    entries = None
                if entries:
                    via = "llm"
            if entries is None:
                entries = _pick_deterministic(project_id, genre, seed)
            html = _assemble(entries)
            if len(html.encode("utf-8")) > _MAX_HTML_BYTES:
                raise CompositionError("composed page exceeds 150KB budget")
        except Exception:
            _log.exception("composition failed project=%s seed=%s", project_id, seed)
            continue
        _log.info(
            "composed project=%s seed=%s via=%s ms=%.1f bytes=%d",
            project_id,
            seed,
            via,
            (time.perf_counter() - started) * 1000,
            len(html),
        )
        variants.append(
            {"html": html, "componentIds": [e["id"] for e in entries], "via": via}
        )
    return variants

File: src/server/services/compositionService.ts
/** Compose website variants from the extracted component library (port of composition_service.py). */

import fs from "node:fs";
import path from "node:path";
import { backendPath } from "../paths";
import { loadOnboardingProvider } from "../llmProvider";

const COMPONENTS_DIR = backendPath("studio/components");
const TEMPLATES_DIR = backendPath("studio/templates");

const RECIPE = [
  "nav",
  "hero",
  "features",
  "about",
  "gallery",
  "stats",
  "pricing",
  "testimonials",
  "faq",
  "cta",
  "contact",
  "footer",
] as const;
const REQUIRED = new Set(["nav", "hero", "footer"]);
const MAX_HTML_BYTES = 150 * 1024;

interface ComponentEntry {
  id: string;
  type: string;
  genre: string;
  tags: string[];
  path: string;
  fonts?: string[];
}

let indexCache: Record<string, ComponentEntry[]> | null = null;
const fileCache = new Map<string, string>();

const SHELL = `<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>{{BRAND}}</title>
<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/css/bootstrap.min.css" rel="stylesheet">
__FONTS__
<style>__BASE__</style>
<style>__PALETTE__</style>
</head>
<body>
__BODY__
</body>
</html>`;

const SELECT_PROMPT = `Pick the page design that best fits this small business.

Business genre: {genre}
Business:
{facts}

Page designs (JSON) — each is one complete, visually coherent page:
{designs}

Reply with only JSON naming ONE design, like {"design": "2"}.
Use only a design id listed above.`;

export class CompositionError extends Error {}

export function resetForTests(): void {
  indexCache = null;
  fileCache.clear();
}

function index(): Record<string, ComponentEntry[]> {
  if (indexCache) return indexCache;
  const raw = fs.readFileSync(path.join(COMPONENTS_DIR, "manifest.json"), "utf8");
  const entries = JSON.parse(raw) as ComponentEntry[];
  const idx: Record<string, ComponentEntry[]> = {};
  for (const entry of entries) {
    (idx[entry.type] ??= []).push(entry);
  }
  indexCache = idx;
  return idx;
}

function read(name: string): string {
  if (!fileCache.has(name)) {
    fileCache.set(name, fs.readFileSync(path.join(COMPONENTS_DIR, name), "utf8"));
  }
  return fileCache.get(name)!;
}

function family(entry: ComponentEntry): string {
  const parts = entry.id.split("-");
  return parts.length === 4 ? parts[1]! : "0";
}

function order(entry: ComponentEntry): number {
  const parts = entry.id.split("-");
  return parts.length === 4 ? parseInt(parts[2]!, 10) : parseInt(parts[1]!, 10);
}

function families(genre: string): Record<string, Record<string, ComponentEntry[]>> {
  const fams: Record<string, Record<string, ComponentEntry[]>> = {};
  for (const slot of RECIPE) {
    for (const entry of index()[slot] ?? []) {
      if (entry.genre !== genre) continue;
      const fam = family(entry);
      (fams[fam] ??= {})[slot] ??= [];
      fams[fam]![slot]!.push(entry);
    }
  }
  const out: Record<string, Record<string, ComponentEntry[]>> = {};
  for (const [fam, slots] of Object.entries(fams)) {
    if ([...REQUIRED].every((s) => slots[s])) out[fam] = slots;
  }
  return out;
}

const ROOT_RE = /:root\s*\{[\s\S]*?\}/;

function palette(entry: ComponentEntry): string {
  const fam = family(entry);
  const name = fam === "0" ? `${entry.genre}.html` : `${entry.genre}-${fam}.html`;
  if (!fileCache.has(name)) {
    try {
      const found = ROOT_RE.exec(fs.readFileSync(path.join(TEMPLATES_DIR, name), "utf8"));
      fileCache.set(name, found?.[0] ?? "");
    } catch {
      fileCache.set(name, "");
    }
  }
  return fileCache.get(name)!;
}

function mulberry32(seed: number): () => number {
  return () => {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function seededRandom(seed: string): { next: () => number; choice: <T>(items: T[]) => T } {
  let h = 0;
  for (let i = 0; i < seed.length; i++) h = (Math.imul(31, h) + seed.charCodeAt(i)) | 0;
  const rnd = mulberry32(h);
  return {
    next: rnd,
    choice: <T>(items: T[]) => items[Math.floor(rnd() * items.length)]!,
  };
}

function build(slots: Record<string, ComponentEntry[]>, rng: ReturnType<typeof seededRandom>): ComponentEntry[] {
  const chosen = RECIPE.filter((s) => slots[s]).map((s) => rng.choice(slots[s]!));
  return chosen.sort((a, b) => order(a) - order(b));
}

function pickDeterministic(projectId: string, genre: string, seed: number): ComponentEntry[] {
  const fams = families(genre);
  const keys = Object.keys(fams).sort();
  if (!keys.length) throw new CompositionError(`no complete ${genre} component family`);
  const rng = seededRandom(`${projectId}:${seed}`);
  return build(fams[rng.choice(keys)]!, rng);
}

async function pickLlm(
  genre: string,
  businessFacts: string,
  rng: ReturnType<typeof seededRandom>,
): Promise<ComponentEntry[] | null> {
  const [provider, model] = loadOnboardingProvider();
  if (!provider || !model) return null;
  const fams = families(genre);
  const keys = Object.keys(fams);
  if (!keys.length) return null;
  const designs: Record<string, unknown> = {};
  for (const f of keys) {
    const slots = fams[f]!;
    designs[f] = {
      sections: RECIPE.filter((s) => slots[s]),
      tags: [
        ...new Set(
          Object.values(slots)
            .flat()
            .flatMap((e) => e.tags)
            .filter((t) => t !== genre),
        ),
      ].sort(),
    };
  }
  const prompt = SELECT_PROMPT.replace("{genre}", genre)
    .replace("{facts}", businessFacts || "(no details provided)")
    .replace("{designs}", JSON.stringify(designs, null, 0));
  try {
    const out = await provider.generate(model, prompt, { numCtx: 8192, timeoutS: 30 });
    const raw = (out ?? "").trim().replace(/^```(?:json)?\s*|\s*```$/g, "");
    const data = JSON.parse(raw) as { design?: string };
    const slots = fams[String(data.design ?? "")];
    if (!slots) return null;
    return build(slots, rng);
  } catch {
    return null;
  }
}

function assemble(entries: ComponentEntry[]): string {
  const fonts: string[] = [];
  const seen = new Set<string>();
  const parts: string[] = [];
  for (const entry of entries) {
    parts.push(read(entry.path));
    for (const href of entry.fonts ?? []) {
      if (!seen.has(href)) {
        seen.add(href);
        fonts.push(`<link href="${href}" rel="stylesheet">`);
      }
    }
  }
  const html = SHELL.replace("__FONTS__", fonts.join("\n"))
    .replace("__BASE__", read("base.css"))
    .replace("__PALETTE__", palette(entries[0]!))
    .replace("__BODY__", parts.join("\n"));
  for (const slot of ["nav", "hero", "footer"] as const) {
    if (!html.includes(`data-section="${slot}"`)) {
      throw new CompositionError(`assembled html missing data-section=${slot}`);
    }
  }
  return html;
}

export interface ComposedVariant {
  html: string;
  componentIds: string[];
  via: string;
}

export async function composeProjectVariants(
  projectId: string,
  businessFacts: string,
  genre: string,
  count: number,
): Promise<ComposedVariant[]> {
  if (count <= 0) return [];
  const variants: ComposedVariant[] = [];
  for (let seed = 0; seed < count; seed++) {
    try {
      let entries: ComponentEntry[] | null = null;
      let via = "deterministic";
      if (seed === 0) {
        try {
          entries = await pickLlm(genre, businessFacts, seededRandom(`${projectId}:llm`));
        } catch {
          entries = null;
        }
        if (entries) via = "llm";
      }
      if (!entries) entries = pickDeterministic(projectId, genre, seed);
      const html = assemble(entries);
      if (Buffer.byteLength(html, "utf8") > MAX_HTML_BYTES) {
        throw new CompositionError("composed page exceeds 150KB budget");
      }
      variants.push({
        html,
        componentIds: entries.map((e) => e.id),
        via,
      });
    } catch {
      continue;
    }
  }
  return variants;
}
